File: android/src/main/java/com/margelo/nitro/obitrain/reactnativenotifications/ReactNativeNotifications.kt
package com.margelo.nitro.obitrain.reactnativenotifications

import android.app.Activity
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.ActivityEventListener
import com.google.firebase.messaging.FirebaseMessaging
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import org.json.JSONObject

@DoNotStrip
class ReactNativeNotifications : HybridReactNativeNotificationsSpec() {
  private var registrationFailedCallback: ((RegistrationErrorEvent) -> Unit)? = null
  private val mainHandler = Handler(Looper.getMainLooper())

  private val context: Context?
    get() = NitroModules.applicationContext

  init {
    // warm taps: the tray intent re-enters the (singleTask) activity through onNewIntent
    NitroModules.applicationContext?.addActivityEventListener(object : ActivityEventListener {
      override fun onNewIntent(intent: Intent) {
        val extras = intent.extras ?: return
        if (NotificationPayload.isNotificationExtras(extras)) {
          Log.d(TAG, "notification opened via onNewIntent")
          NotificationEvents.emitOpened(NotificationPayload.fromExtras(extras).toString())
        }
      }

      override fun onActivityResult(
        activity: Activity, requestCode: Int, resultCode: Int, data: Intent?
      ) = Unit
    })
  }

  override fun registerRemoteNotifications(): Promise<Unit> {
    val promise = Promise<Unit>()
    fetchToken(promise, 0)
    return promise
  }

  @Suppress("DEPRECATION")
  private fun fetchToken(promise: Promise<Unit>, retryCount: Int) {
    FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
      if (task.isSuccessful) {
        NotificationEvents.emitToken(task.result)
        promise.resolve(Unit)
        return@addOnCompleteListener
      }
      if (retryCount < TOKEN_FETCH_MAX_RETRIES) {
        val retryDelay = TOKEN_FETCH_RETRY_DELAY_MS * (1L shl retryCount)
        mainHandler.postDelayed(
          { fetchToken(promise, retryCount + 1) },
          retryDelay
        )
        return@addOnCompleteListener
      }
      registrationFailedCallback?.invoke(
        RegistrationErrorEvent(
          code = "fcm-token-fetch-failed",
          domain = "com.google.firebase.messaging",
          localizedDescription = task.exception?.message ?: "Unknown FCM token error"
        )
      )
      promise.resolve(Unit)
    }
  }

  override fun getInitialNotification(): Promise<String?> {
    // our own tray entries go through the trampoline; FCM-rendered ones land in the
    // launcher activity's intent extras
    NotificationEvents.initialNotification?.let {
      NotificationEvents.initialNotification = null
      return Promise.resolved(it)
    }
    val extras = NitroModules.applicationContext?.currentActivity?.intent?.extras
    val json = extras
      ?.takeIf { NotificationPayload.isNotificationExtras(it) }
      ?.let { NotificationPayload.fromExtras(it).toString() }
    return Promise.resolved(json)
  }

  override fun postLocalNotification(payloadJson: String) {
    val ctx = context ?: return
    val payload = runCatching { JSONObject(payloadJson) }.getOrNull() ?: JSONObject()
    NotificationDisplay.post(ctx, payload)
  }

  override fun setNotificationChannel(channel: NotificationChannelConfig) {
    val ctx = context ?: return
    val importance = channel.importance?.toInt() ?: NotificationManager.IMPORTANCE_DEFAULT
    val nativeChannel = NotificationChannel(channel.channelId, channel.name, importance).apply {
      channel.description?.let { description = it }
      channel.enableLights?.let { enableLights(it) }
      channel.enableVibration?.let { enableVibration(it) }
      channel.showBadge?.let { setShowBadge(it) }
      channel.vibrationPattern?.let { pattern ->
        vibrationPattern = pattern.map { it.toLong() }.toLongArray()
      }
    }
    val manager = ctx.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    manager.createNotificationChannel(nativeChannel)
  }

  override fun onTokenReceived(callback: (String) -> Unit) {
    NotificationEvents.setTokenCallback(callback)
  }

  override fun onRegistrationFailed(callback: (RegistrationErrorEvent) -> Unit) {
    registrationFailedCallback = callback
  }

  // Android has no OS completion block to honour, so the JS promises are awaited only to
  // surface rejections instead of leaking them.
  override fun onNotificationReceivedForeground(
    callback: (String) -> Promise<Promise<NotificationCompletion>>
  ) {
    NotificationEvents.setForegroundCallback { json -> callback(json).settle() }
  }

  override fun onNotificationOpened(callback: (String) -> Promise<Promise<Unit>>) {
    NotificationEvents.setOpenedCallback { json -> callback(json).settle() }
  }

  override fun onNotificationReceivedBackground(
    callback: (String) -> Promise<Promise<BackgroundFetchResult>>
  ) {
    NotificationEvents.setBackgroundCallback { json -> callback(json).settle() }
  }

  private companion object {
    const val TAG = "ObiNotifications"
    const val TOKEN_FETCH_MAX_RETRIES = 3
    const val TOKEN_FETCH_RETRY_DELAY_MS = 10_000L
  }

  private fun <T> Promise<Promise<T>>.settle() {
    then { inner -> inner.catch { } }.catch { }
  }
}

File: ios/nitro/ReactNativeNotifications.swift
import Foundation
import NitroModules
import ObiNotificationsCore
import UIKit
import UserNotifications

/** Nitro bridge: adapts JS callbacks/promises onto ObiNotificationsCenter. */
class ReactNativeNotifications: HybridReactNativeNotificationsSpec {
  private var center: ObiNotificationsCenter { ObiNotificationsCenter.shared }

  func registerRemoteNotifications() throws -> Promise<Void> {
    let promise = Promise<Void>()
    UNUserNotificationCenter.current().requestAuthorization(options: [
      .alert, .badge, .sound,
    ]) { granted, error in
      if let error {
        ObiNotificationsCenter.didFailToRegisterForRemoteNotifications(error: error)
        promise.resolve()
        return
      }
      if granted {
        DispatchQueue.main.async {
          UIApplication.shared.registerForRemoteNotifications()
        }
      }
      promise.resolve()
    }
    return promise
  }

  func getInitialNotification() throws -> Promise<String?> {
    let json = center.initialNotificationJson
    center.initialNotificationJson = nil
    return Promise.resolved(withResult: json)
  }

  func postLocalNotification(payloadJson: String) throws {
    guard let data = payloadJson.data(using: .utf8),
      let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { throw NSError(domain: "ObiNotifications", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid local notification payload"]) }
    let content = UNMutableNotificationContent()
    content.title = payload["title"] as? String ?? ""
    content.body = payload["body"] as? String ?? ""
    content.sound = .default
    content.userInfo = payload
    let request = UNNotificationRequest(
      identifier: payload["id"] as? String ?? UUID().uuidString,
      content: content,
      trigger: UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false))
    let notifications = UNUserNotificationCenter.current()
    notifications.requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
      if let error { NSLog("Local notification permission failed: %@", error.localizedDescription); return }
      guard granted else { return }
      notifications.add(request) { error in
        if let error { NSLog("Local notification failed: %@", error.localizedDescription) }
      }
    }
  }

  func setNotificationChannel(channel: NotificationChannelConfig) throws {
    // Android-only API.
  }

  func onTokenReceived(callback: @escaping (String) -> Void) throws {
    center.tokenHandler = callback
    // replay so registration order doesn't matter
    if let token = center.lastToken {
      callback(token)
    }
  }

  func onRegistrationFailed(callback: @escaping (RegistrationErrorEvent) -> Void) throws {
    center.registrationFailedHandler = { code, domain, message in
      callback(
        RegistrationErrorEvent(code: code, domain: domain, localizedDescription: message))
    }
  }

  func onNotificationReceivedForeground(
    callback: @escaping (String) -> Promise<Promise<NotificationCompletion>>
  ) throws {
    center.foregroundHandler = { json, completion in
      callback(json).then { inner in
        inner.then { result in
          completion(result.alert == true, result.sound == true, result.badge == true)
        }.catch { _ in completion(false, false, false) }
      }.catch { _ in completion(false, false, false) }
    }
  }

  func onNotificationOpened(callback: @escaping (String) -> Promise<Promise<Void>>) throws {
    center.openedHandler = { json, completion in
      callback(json).then { inner in
        inner.then { completion() }.catch { _ in completion() }
      }.catch { _ in completion() }
    }
  }

  func onNotificationReceivedBackground(
    callback: @escaping (String) -> Promise<Promise<BackgroundFetchResult>>
  ) throws {
    center.backgroundHandler = { json, completion in
      callback(json).then { inner in
        inner.then { result in
          switch result {
          case .newdata: completion("newData")
          case .nodata: completion("noData")
          case .failed: completion("failed")
          }
        }.catch { _ in completion("failed") }
      }.catch { _ in completion("failed") }
    }
  }
}
